File: pickem-api/app/schemas/standing.py
import uuid
from datetime import datetime

from pydantic import BaseModel, computed_field


class StandingRead(BaseModel):
    user_id: uuid.UUID
    group_id: uuid.UUID
    display_name: str
    wins: int
    losses: int
    superdog_wins: int
    superdogs_used: int
    updated_at: datetime

    model_config = {"from_attributes": True}

    @computed_field  # type: ignore[prop-decorator]
    @property
    def win_percentage(self) -> float:
        # Each superdog win counts as 3 regular wins in the record.
        effective_wins = self.wins + self.superdog_wins * 3
        total = effective_wins + self.losses
        return effective_wins / total if total > 0 else 0.0

    @computed_field  # type: ignore[prop-decorator]
    @property
    def record(self) -> str:
        effective_wins = self.wins + self.superdog_wins * 3
        return f"{effective_wins}-{self.losses}"

File: PickEm-iOS/PickEm/Domain/Models/Standing.swift
import Foundation

struct Standing: Identifiable, Equatable, Hashable, Codable {
    let userID: String
    let groupID: String
    let displayName: String
    let wins: Int
    let losses: Int
    let superdogWins: Int
    let superdogsUsed: Int
    let updatedAt: Date

    var id: String { "\(userID)-\(groupID)" }

    /// A superdog win counts as 3 regular wins — matches the backend's
    /// StandingRead.win_percentage/record computation
    /// (pickem-api/app/schemas/standing.py) and the same weighting
    /// PlayerHistoryRow already used for its own running record.
    var effectiveWins: Int { wins + superdogWins * 3 }

    var totalGames: Int { effectiveWins + losses }

    var winPercentage: Double {
        guard totalGames > 0 else { return 0 }
        return Double(effectiveWins) / Double(totalGames)
    }

    var record: String { "\(effectiveWins)-\(losses)" }
}
